lightning-net/src/lib.rs

```diff
@@ -0,0 +1,1146 @@
+        let mut buf = [0; 8192];
+
+        loop {
+            // Every time this line is reached, read_paused == false.
+            // Do a non-blocking try_recv() to see if we've been asked to pause reads
+            let shutdown = self.handle_command(false);
+            if shutdown {
+                break;
+            }
+
+            if self.read_paused {
+                // To avoid a busy loop while reading is paused, block on the
+                // reader_cmd channel until we are told to resume reading again
+                // or until we receive a shut down command.
+                let shutdown = self.handle_command(true);
+                if shutdown {
+                    break;
+                }
+            } else {
+                // Reading is not paused; block on the next read.
+                // If the SyncSocketDescriptor disconnects the underlying TcpStream,
+                // the Reader will read Ok(0), in which case we know to break
+                // the loop and shut down.
+                match self.inner.read(&mut buf) {
+                    Ok(0) => {
+                        // Peer disconnected. Notify the PeerManager then break
+                        self.peer_manager.socket_disconnected(&self.descriptor);
+                        self.peer_manager.process_events();
+                        break;
+                    }
+                    Ok(bytes_read) => {
+                        // Register the read event with the PeerManager
+                        match self
+                            .peer_manager
+                            .read_event(&mut self.descriptor, &buf[0..bytes_read])
+                        {
+                            Ok(pause_read) => {
+                                if pause_read {
+                                    self.read_paused = true;
+                                }
+                            }
+                            Err(_) => {
+                                // Rust-Lightning told us to disconnect; do it
+                                // No need to notify PeerManager in this case
+                                break;
+                            }
+                        }
+
+                        // As noted in the read_event() docs, call process_events().
+                        self.peer_manager.process_events()
+                    }
+                    Err(e) => match e.kind() {
+                        TimedOut | Interrupted => {
+                            // Retry
+                        }
+                        _ => {
+                            // For all other errors, notify the
+                            // PeerManager, break, and shut down
+                            self.peer_manager.socket_disconnected(&self.descriptor);
+                            self.peer_manager.process_events();
+                            break;
+                        }
+                    },
+                }
+            }
+        }
```

Comment:
Implemented in 98fdb932879d0eec765375ebb65311d79572d61f

The only thing I could really do is re-duplicate `handle_command()` into `do_recv()` and `do_try_recv()` which don't take a `bool` parameter. Feels pretty dumb since `recv()` and `try_recv()` are handled the exact same way, except that `recv()` returns a `Result<_, RecvError>` while `try_recv()` returns a `Result<_, TryRecvError>`. Any suggestions?

Other than extracting `do_read()` I think the old flow looks better tbh